File: app/src/main/java/com/oxclient/core/proxy/PacketProcessor.java
package com.oxclient.core.proxy;

import android.util.Log;

import com.oxclient.events.PacketEvent;
import com.oxclient.events.PacketEventBus;
import com.oxclient.session.SessionManager;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.net.InetSocketAddress;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.zip.DeflaterOutputStream;
import java.util.zip.InflaterInputStream;

/**
 * PacketProcessor
 *
 * Bedrock Batch (0xFE 0x01) paketlerini açar → sub-paketleri dispatch eder → re-pack eder.
 *
 * DÜZELTME 1 — Batch decompression:
 *   Önceki kod batch'i açmadan 0xFE olarak dispatch ediyordu.
 *   Hiçbir modül BATCH event'i dinlemediğinden StartGame/AddPlayer/MovePlayer
 *   hiç işlenmiyordu → EntityTracker boş kalıyordu → KillAura hedef bulamıyordu.
 *
 * DÜZELTME 2 — EntityTracker.onClientPacket():
 *   C2S paketleri dispatch edildikten sonra EntityTracker'a bildirilir.
 *   Bu sayede PLAYER_AUTH_INPUT paketi selfX/Y/Z günceller.
 *   Önceki kodda bu çağrı hiç yoktu → selfX=0,selfY=0,selfZ=0 → tüm entityler
 *   "range dışında" → KillAura hiç saldırmıyordu.
 */
public class PacketProcessor {
    private static final String TAG = "PacketProcessor";

    // Batch packet id VarInt = 0xFE → 2-byte encoding: 0xFE 0x01
    private static final int BATCH_VARINT_B0 = 0xFE;
    private static final int BATCH_VARINT_B1 = 0x01;

    private static final Map<Integer, String> NAMES = new HashMap<>();
    static {
        NAMES.put(PacketIds.LOGIN,                "Login");
        NAMES.put(PacketIds.PLAY_STATUS,          "PlayStatus");
        NAMES.put(PacketIds.START_GAME,           "StartGame");
        NAMES.put(PacketIds.ADD_PLAYER,           "AddPlayer");
        NAMES.put(PacketIds.ADD_ENTITY,           "AddEntity");
        NAMES.put(PacketIds.REMOVE_ENTITY,        "RemoveEntity");
        NAMES.put(PacketIds.MOVE_PLAYER,          "MovePlayer");
        NAMES.put(PacketIds.MOVE_ENTITY_ABSOLUTE, "MoveEntityAbs");
        NAMES.put(PacketIds.MOVE_ENTITY_DELTA,    "MoveEntityDelta");
        NAMES.put(PacketIds.PLAYER_AUTH_INPUT,    "PlayerAuthInput");
        NAMES.put(PacketIds.INVENTORY_TRANSACTION,"InventoryTransaction");
        NAMES.put(PacketIds.INVENTORY_CONTENT,    "InventoryContent");
        NAMES.put(PacketIds.INVENTORY_SLOT,       "InventorySlot");
        NAMES.put(PacketIds.SET_ENTITY_DATA,      "SetEntityData");
        NAMES.put(PacketIds.UPDATE_ATTRIBUTES,    "UpdateAttributes");
        NAMES.put(PacketIds.ANIMATE,              "Animate");
        NAMES.put(PacketIds.RESPAWN,              "Respawn");
        NAMES.put(PacketIds.TEXT,                 "Text");
        NAMES.put(PacketIds.BATCH,                "Batch");
        NAMES.put(PacketIds.DISCONNECT,           "Disconnect");
        NAMES.put(PacketIds.NETWORK_STACK_LATENCY,"NetworkLatency");
    }

    // ── Public API ────────────────────────────────────────────────────────

    public byte[] processC2S(byte[] payload, InetSocketAddress addr) {
        if (payload == null || payload.length == 0) return payload;
        if (isBatch(payload)) return processBatch(payload, addr, PacketEvent.DIRECTION_C2S);
        return dispatchAndTrack(payload, addr, PacketEvent.DIRECTION_C2S);
    }

    public byte[] processS2C(byte[] payload, InetSocketAddress addr) {
        if (payload == null || payload.length == 0) return payload;
        if (isBatch(payload)) return processBatch(payload, addr, PacketEvent.DIRECTION_S2C);
        return dispatchAndTrack(payload, addr, PacketEvent.DIRECTION_S2C);
    }

    // ── Batch ─────────────────────────────────────────────────────────────

    /** Bedrock Batch: ilk 2 byte = VarInt 0xFE (0xFE 0x01) */
    private boolean isBatch(byte[] p) {
        return p.length > 2
            && (p[0] & 0xFF) == BATCH_VARINT_B0
            && (p[1] & 0xFF) == BATCH_VARINT_B1;
    }

    private byte[] processBatch(byte[] payload, InetSocketAddress addr, int dir) {
        try {
            // Header atla (2 byte: 0xFE 0x01)
            byte[] compressed = new byte[payload.length - 2];
            System.arraycopy(payload, 2, compressed, 0, compressed.length);

            byte[]       raw        = inflate(compressed);
            List<byte[]> subPackets = splitSubPackets(raw);
            if (subPackets.isEmpty()) return payload;

            boolean      modified  = false;
            List<byte[]> processed = new ArrayList<>(subPackets.size());

            for (byte[] sub : subPackets) {
                byte[] result = dispatchAndTrack(sub, addr, dir);
                if (result == null) { modified = true; continue; } // dropped
                if (result != sub)    modified = true;
                processed.add(result);
            }

            if (!modified)          return payload;
            if (processed.isEmpty()) return null;   // tümü drop → paketi yutma
            return packBatch(processed);

        } catch (Exception e) {
            Log.e(TAG, "processBatch error: " + e.getMessage());
            return payload;
        }
    }

    private List<byte[]> splitSubPackets(byte[] data) {
        List<byte[]> list = new ArrayList<>();
        ByteBuffer b = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN);
        while (b.hasRemaining()) {
            int len = readVarInt(b);
            if (len <= 0 || len > b.remaining()) break;
            byte[] sub = new byte[len];
            b.get(sub);
            list.add(sub);
        }
        return list;
    }

    private byte[] packBatch(List<byte[]> subs) throws Exception {
        ByteArrayOutputStream rawOut = new ByteArrayOutputStream();
        for (byte[] s : subs) { writeVarInt(rawOut, s.length); rawOut.write(s); }
        byte[] compressed = deflate(rawOut.toByteArray());
        ByteArrayOutputStream out = new ByteArrayOutputStream(compressed.length + 2);
        out.write(BATCH_VARINT_B0);
        out.write(BATCH_VARINT_B1);
        out.write(compressed);
        return out.toByteArray();
    }

    // ── Single packet dispatch ────────────────────────────────────────────

    /**
     * Tek paketi dispatch eder ve EntityTracker'ı günceller.
     * null → paket drop edildi.
     */
    private byte[] dispatchAndTrack(byte[] payload, InetSocketAddress addr, int dir) {
        if (payload == null || payload.length == 0) return payload;

        ByteBuffer buf = ByteBuffer.wrap(payload).order(ByteOrder.LITTLE_ENDIAN);
        int    id   = readVarInt(buf);
        String name = NAMES.getOrDefault(id, String.format("Unk(0x%02X)", id));

        PacketEvent ev = new PacketEvent(id, name, payload, dir, addr);
        PacketEventBus.publish(ev);

        if (ev.isCancelled()) {
            Log.d(TAG, "Dropped: " + name + " dir=" + (dir == 0 ? "C2S" : "S2C"));
            return null;
        }

        // FIX: C2S paketlerini EntityTracker'a ilet (selfX/Y/Z güncelleme)
        if (dir == PacketEvent.DIRECTION_C2S) {
            try {
                SessionManager.INSTANCE.getEntityTracker().onClientPacket(ev);
            } catch (Exception ignored) {}
        }

        return ev.getPayload();
    }

    // ── Zlib ─────────────────────────────────────────────────────────────

    private static byte[] inflate(byte[] data) throws Exception {
        try (InflaterInputStream inf = new InflaterInputStream(new ByteArrayInputStream(data));
             ByteArrayOutputStream out = new ByteArrayOutputStream()) {
            byte[] buf = new byte[8192];
            int n;
            while ((n = inf.read(buf)) != -1) out.write(buf, 0, n);
            return out.toByteArray();
        }
    }

    private static byte[] deflate(byte[] data) throws Exception {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        try (DeflaterOutputStream def = new DeflaterOutputStream(out)) {
            def.write(data);
        }
        return out.toByteArray();
    }

    // ── VarInt ───────────────────────────────────────────────────────────

    public static int readVarInt(ByteBuffer b) {
        int v = 0, s = 0;
        byte c;
        do {
            if (!b.hasRemaining()) break;
            c = b.get();
            v |= (c & 0x7F) << s;
            s += 7;
        } while ((c & 0x80) != 0 && s < 35);
        return v;
    }

    private static void writeVarInt(ByteArrayOutputStream out, int v) {
        do {
            byte c = (byte)(v & 0x7F);
            v >>>= 7;
            if (v != 0) c |= (byte) 0x80;
            out.write(c);
        } while (v != 0);
    }
}

File: app/src/main/kotlin/com/oxclient/core/proxy/PacketProcessor.kt
package com.oxclient.core.proxy

import android.util.Log
import com.oxclient.events.PacketEvent
import com.oxclient.events.PacketEventBus
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream

/**
 * PacketProcessor
 *
 * Diğer AI'nın verdiği "otomatik sıkıştırma tespiti" kodu 2 nedenden yanlış:
 *
 * 1) firstByte == 0x78 ise zlib aktif ediyordu.
 *    0x78 ham batch'te sıradan bir varint length byte'ı olabilir (120 byte'lık paket).
 *    Bu heuristic yanlış pozitif üretir → ham paket inflate edilmeye çalışılır
 *    → exception → fallback yine compression açar → sonsuz döngü / bağlantı kopması.
 *
 * 2) NetworkSettings algoritma okuma offset'i hatalıydı.
 *    NetworkSettings (0xC7) formatı:
 *      [packetId varint=2B] [threshold uint16LE=2B] [algorithm uint16LE=2B]
 *    Algoritmanın düşük byte'ı: data[idSize + 2]
 *    Önceki kodda idSize+2 ve idSize+3 LE birleştirme yapıyordu — gereksiz ve
 *    büyük değerlerde yanlış sonuç verebilirdi.
 *
 * DOĞRU: Sıkıştırma durumu YALNIZCA NetworkSettings paketinden okunur.
 *
 * ✅ FIX (KRİTİK): Bedrock şifreleme (AES-256-CFB8) desteği eklendi.
 *    2b2t gibi online sunucularda ServerToClientHandshake (0x03) paketi
 *    geldikten sonra TÜM paketler şifreli gelir.
 *    Şifre çözülmeden PacketProcessor ham byte okur → hiçbir şey parse edilemez
 *    → selfRuntimeId=0, konum=0,0,0 → modüller çalışmaz.
 */
object PacketProcessor {

    private const val TAG = "PacketProcessor"

    @Volatile var compressionEnabled   = false
    @Volatile var compressionAlgorithm = 0     // 0 = zlib, 1 = snappy

    // ✅ FIX: AES-256-CFB8 şifreleme state'i
    // Bedrock şifrelemesi: her yön için ayrı cipher, counter mod (send counter / receive counter)
    @Volatile var encryptionEnabled    = false
    private var decryptCipher: javax.crypto.Cipher? = null  // sunucu→biz
    private var encryptCipher: javax.crypto.Cipher? = null  // biz→sunucu
    @Volatile private var sendCounter  = 0L
    @Volatile private var recvCounter  = 0L
    private lateinit var secretKey: javax.crypto.SecretKey

    /**
     * ServerToClientHandshake JWT'sinden şifreli oturumu başlat.
     * WClient'in OnlineLoginPacketListener.beforeServerBound içinde yaptığının aynısı.
     *
     * jwt: ServerToClientHandshake paketinin içindeki JWT string
     * clientPrivateKey: Login sırasında oluşturulan EC private key (Base64)
     */
    fun enableEncryption(secretKeyBytes: ByteArray) {
        try {
            secretKey = javax.crypto.spec.SecretKeySpec(secretKeyBytes, "AES")
            // Bedrock CFB8: IV = secretKey'in ilk 16 byte'ı
            val iv = secretKeyBytes.copyOf(16)
            val ivSpec = javax.crypto.spec.IvParameterSpec(iv)

            decryptCipher = javax.crypto.Cipher.getInstance("AES/CFB8/NoPadding").apply {
                init(javax.crypto.Cipher.DECRYPT_MODE, secretKey, ivSpec)
            }
            encryptCipher = javax.crypto.Cipher.getInstance("AES/CFB8/NoPadding").apply {
                init(javax.crypto.Cipher.ENCRYPT_MODE, secretKey, ivSpec)
            }
            sendCounter = 0L
            recvCounter = 0L
            encryptionEnabled = true
            Log.i(TAG, "✅ Şifreleme aktif (AES-256-CFB8)")
        } catch (e: Exception) {
            Log.e(TAG, "Şifreleme başlatma hatası: ${e.message}")
        }
    }

    fun resetEncryption() {
        encryptionEnabled = false
        decryptCipher = null
        encryptCipher = null
        sendCounter = 0L
        recvCounter = 0L
    }

    fun reset() {
        compressionEnabled = false
        compressionAlgorithm = 0
        resetEncryption()
    }

    /** Sunucudan gelen veriyi çöz (S→C) */
    private fun decrypt(data: ByteArray): ByteArray {
        return decryptCipher?.doFinal(data) ?: data
    }

    /** Sunucuya gönderilecek veriyi şifrele (C→S inject) */
    fun encrypt(data: ByteArray): ByteArray {
        return encryptCipher?.doFinal(data) ?: data
    }

    fun processBatch(body: ByteArray, direction: PacketEvent.Direction): ByteArray? {
        if (body.isEmpty()) return body
        return try {
            // ✅ FIX: Sunucudan gelen veri şifrelenmiş olabilir — önce çöz
            val decrypted = if (encryptionEnabled && direction == PacketEvent.Direction.SERVER_TO_CLIENT) {
                decrypt(body)
            } else {
                body
            }
            val result = if (compressionEnabled) processCompressedBatch(decrypted, direction)
                         else                    processRawBatch(decrypted, direction)
            // ✅ FIX: Sunucuya enjekte edilecek paketleri şifrele
            if (encryptionEnabled && direction == PacketEvent.Direction.CLIENT_TO_SERVER && result != null) {
                encrypt(result)
            } else {
                result
            }
        } catch (e: Exception) {
            Log.w(TAG, "Batch işleme hatası, orijinal iletiliyor: ${e.message}")
            body
        }
    }

    private fun processRawBatch(body: ByteArray, direction: PacketEvent.Direction): ByteArray? {
        val stream  = ByteArrayInputStream(body)
        val packets = mutableListOf<ByteArray>()
        var changed = false

        while (stream.available() > 0) {
            val len = readVarInt(stream)
            if (len <= 0 || stream.available() < len) break
            val data = ByteArray(len)
            stream.read(data)
            val result = processSinglePacket(data, direction)
            when {
                result == null              -> { changed = true }
                !result.contentEquals(data) -> { changed = true; packets.add(result) }
                else                        -> packets.add(data)
            }
        }

        if (packets.isEmpty()) return null
        if (!changed) return body
        val out = ByteArrayOutputStream(body.size)
        for (pkt in packets) { writeVarInt(out, pkt.size); out.write(pkt) }
        return out.toByteArray()
    }

    private fun processCompressedBatch(body: ByteArray, direction: PacketEvent.Direction): ByteArray? {
        return try {
            val decompressed = when (compressionAlgorithm) {
                1    -> decompressSnappy(body)
                else -> zlibInflate(body)
            }
            val processed = processRawBatch(decompressed, direction) ?: return null
            when (compressionAlgorithm) {
                1    -> compressSnappy(processed)
                else -> zlibDeflate(processed)
            }
        } catch (e: Exception) {
            Log.w(TAG, "Sıkıştırma hatası: ${e.message}")
            body
        }
    }

    private fun zlibInflate(input: ByteArray): ByteArray {
        val inf = java.util.zip.Inflater(); inf.setInput(input)
        val out = ByteArrayOutputStream(); val buf = ByteArray(4096)
        while (!inf.finished()) { val n = inf.inflate(buf); if (n == 0) break; out.write(buf, 0, n) }
        inf.end(); return out.toByteArray()
    }

    private fun zlibDeflate(input: ByteArray): ByteArray {
        val def = java.util.zip.Deflater(java.util.zip.Deflater.DEFAULT_COMPRESSION)
        def.setInput(input); def.finish()
        val out = ByteArrayOutputStream(); val buf = ByteArray(4096)
        while (!def.finished()) { val n = def.deflate(buf); if (n > 0) out.write(buf, 0, n) }
        def.end(); return out.toByteArray()
    }

    private fun decompressSnappy(input: ByteArray): ByteArray = try {
        org.iq80.snappy.Snappy.uncompress(input, 0, input.size)
    } catch (e: Exception) {
        Log.w(TAG, "Snappy başarısız, zlib: ${e.message}"); zlibInflate(input)
    }

    private fun compressSnappy(input: ByteArray): ByteArray = try {
        org.iq80.snappy.Snappy.compress(input)
    } catch (e: Exception) {
        Log.w(TAG, "Snappy compress başarısız, zlib: ${e.message}"); zlibDeflate(input)
    }

    private fun processSinglePacket(data: ByteArray, direction: PacketEvent.Direction): ByteArray? {
        if (data.isEmpty()) return data
        val packetId = readVarInt(ByteArrayInputStream(data))

        // NetworkSettings (0xC7) — sıkıştırmayı aktif et
        // Format: [packetId varint] [threshold uint16LE] [algorithm uint16LE]
        // Sıkıştırma bilgisi SADECE buradan okunur, başka hiçbir heuristik yok.
        if (packetId == 0xC7 && direction == PacketEvent.Direction.SERVER_TO_CLIENT) {
            try {
                val idSize = writeVarInt(packetId).size  // 0xC7 → varint 2 byte
                // algorithm düşük byte'ı: data[idSize + 2]
                compressionAlgorithm = if (data.size >= idSize + 3) {
                    data[idSize + 2].toInt() and 0xFF
                } else 0
                compressionEnabled = true
                Log.i(TAG, "NetworkSettings → ${if (compressionAlgorithm == 1) "Snappy" else "zlib"}")
            } catch (_: Exception) {
                compressionEnabled = true; compressionAlgorithm = 0
            }
        }

        // EntityTracker — HER İKİ YÖN (C→S de dahil, PLAYER_AUTH_INPUT için)
        try { EntityTracker.onPacket(PacketEvent(packetId, data, direction)) } catch (_: Exception) {}

        val event = PacketEvent(packetId, data, direction)
        try { PacketEventBus.publish(event) } catch (e: Exception) {
            Log.e(TAG, "EventBus hatası pkt=0x${packetId.toString(16)}: ${e.message}")
        }

        return when {
            event.isCancelled          -> null
            event.modifiedData != null -> event.modifiedData
            else                       -> data
        }
    }

    fun readVarInt(stream: ByteArrayInputStream): Int {
        var result = 0; var shift = 0
        repeat(5) {
            val b = stream.read()
            if (b == -1) return result
            result = result or ((b and 0x7F) shl shift)
            if (b and 0x80 == 0) return result
            shift += 7
        }
        return result
    }

    fun writeVarInt(out: ByteArrayOutputStream, value: Int) {
        var v = value
        do { var b = v and 0x7F; v = v ushr 7; if (v != 0) b = b or 0x80; out.write(b) } while (v != 0)
    }

    fun writeVarInt(value: Int): ByteArray {
        val out = ByteArrayOutputStream(4); writeVarInt(out, value); return out.toByteArray()
    }
}
